File: leetcode/0547-number-of-provinces/solution.ts
function findCircleNum(isConnected: number[][]): number {
  const n = isConnected.length;
  const seen: boolean[] = Array.from<boolean>({ length: n + 1 }).fill(false);

  function dfs(root: number): void {
    seen[root] = true;

    for (let i = 0; i < n; i++) {
      if (isConnected[root][i] && !seen[i]) {
        dfs(i);
      }
    }
  }

  let ans = 0;

  for (let i = 0; i < n; i++) {
    if (!seen[i]) {
      ans++;
      dfs(i);
    }
  }

  return ans;
}

File: leetcode/0547-number-of-provinces/solution.py
def findCircleNum(self, is_connected: List[List[int]]) -> int:
    n = len(is_connected)
    seen = [False] * n

    def dfs(root: int):
        seen[root] = True

        for i in range(n):
            if is_connected[root][i] and not seen[i]:
                dfs(i)

    ans = 0

    for i in range(n):
        if not seen[i]:
            ans += 1
            dfs(i)

    return ans
